### PR TITLE
No snapshot after start in epoch

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
In this PR: Add a specific key-value to each trie storer. When snapshoting, check if that value is present in the last storage, and only then start the snapshot. This way, after start in epoch no snapshot is triggered.

### CODE REVIEW

File: trie/trieStorageManager.go [hunk 1]
```diff
@@ -613,10 +635,14 @@ func (tsm *trieStorageManager) SetEpochForPutOperation(epoch uint32) {
 	storer.SetEpochForPutOperation(epoch)
 }
 
+// EpochConfirmed is called whenever a new epoch is confirmed
 func (tsm *trieStorageManager) EpochConfirmed(epoch uint32, _ uint64) {
 	tsm.flagDisableOldStorage.Toggle(epoch >= tsm.disableOldStorageEpoch)
 	log.Debug("old trie storage", "disabled", tsm.flagDisableOldStorage.IsSet())
 
+	err := tsm.mainStorer.Put([]byte(activeDBKey), []byte(activeDBVal))
```
Comment (bogdan-rosianu) [inline trie/trieStorageManager.go:643]: it looks like this is the only place when you set the value for active DB. I guess it is meant to be set only on the first confirmed epoch, right? So if starting from genesis, epoch 0 -> epoch 1 will trigger the value to be set; or when starting in epoch, only the next epoch?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: trie/trieStorageManager.go [hunk 2]
```diff
@@ -493,6 +502,19 @@ func (tsm *trieStorageManager) takeCheckpoint(checkpointEntry *snapshotsQueueEnt
 	}
 }
 
+func isActiveDb(stsm *snapshotTrieStorageManager) bool {
+	val, err := stsm.Get([]byte(activeDBKey))
```
Comment (SebastianMarian) [inline trie/trieStorageManager.go:506]: Check the err for nil?

File: trie/trieStorageManager.go [hunk 3]
```diff
@@ -493,6 +502,19 @@ func (tsm *trieStorageManager) takeCheckpoint(checkpointEntry *snapshotsQueueEnt
...
+	if bytes.Equal(val, []byte(activeDBVal)) {
+		return true
+	}
+
+	log.Debug("snapshotTrieStorageManager get",
+		"err", err.Error(),
+		"value", val,
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:513]: Refactored

File: trie/trieStorageManager.go [hunk 4]
```diff
@@ -613,10 +635,14 @@ func (tsm *trieStorageManager) SetEpochForPutOperation(epoch uint32) {
 	storer.SetEpochForPutOperation(epoch)
 }
 
+// EpochConfirmed is called whenever a new epoch is confirmed
 func (tsm *trieStorageManager) EpochConfirmed(epoch uint32, _ uint64) {
 	tsm.flagDisableOldStorage.Toggle(epoch >= tsm.disableOldStorageEpoch)
 	log.Debug("old trie storage", "disabled", tsm.flagDisableOldStorage.IsSet())
 
+	err := tsm.mainStorer.Put([]byte(activeDBKey), []byte(activeDBVal))
+	log.LogIfError(err, "error", "set db as activeDB error")
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:644]: Done.

File: trie/trieStorageManager.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:643]: This func is called every time when an epoch is changed, or when the trieStorageManager is initialized, so `activeDBKey` will be saved in each storer (each epoch a new storer is created)

File: trie/trieStorageManager.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply trie/trieStorageManager.go:506]: Done.

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed. 
image

**Pull request merged**